packages/oauth/src/index.ts

```diff
@@ -129,6 +129,87 @@ export class InstallProvider {
         authResult.botToken = queryResult.bot.token;
         authResult.botId = queryResult.bot.id;
         authResult.botUserId = queryResult.bot.userId;
+
+        // Token Rotation Enabled (Bot Token)
+        if (queryResult.bot.refreshToken !== undefined) {
+          authResult.botRefreshToken = queryResult.bot.refreshToken;
+          authResult.botTokenExpiresAt = queryResult.bot.expiresAt; // utc, seconds
+        }
+      }
+
+      // Token Rotation Enabled (User Token)
+      if (queryResult.user.refreshToken !== undefined) {
+        authResult.userRefreshToken = queryResult.user.refreshToken;
+        authResult.userTokenExpiresAt = queryResult.user.expiresAt; // utc, seconds
+      }
+
+      /*
+      * Token Rotation (Expiry Check + Refresh)
+      * The presence of `(bot|user)TokenExpiresAt` indicates having opted into token rotation.
+      * If the token has expired, or will expire within 2 hours, the token is refreshed and
+      * the `authResult` and `Installation` are updated with the new values.
+      */
+      if (authResult.botRefreshToken !== undefined || authResult.userRefreshToken !== undefined) {
+        const currentUTCSec = Math.floor(Date.now() / 1000); // seconds
+        const tokensToRefresh: string[] = [];
+        let client: WebClient;
+
+        if (authResult.botRefreshToken !== undefined && authResult.botTokenExpiresAt !== undefined) {
+          const botTokenExpiresIn = authResult.botTokenExpiresAt - currentUTCSec;
+          if (botTokenExpiresIn <= 7200) { // 2 hours
+            tokensToRefresh.push(authResult.botRefreshToken);
+          }
+        }
+
+        if (authResult.userRefreshToken !== undefined && authResult.userTokenExpiresAt !== undefined) {
+          const userTokenExpiresIn = authResult.userTokenExpiresAt - currentUTCSec;
+          if (userTokenExpiresIn <= 7200) { // 2 hours
+            tokensToRefresh.push(authResult.userRefreshToken);
+          }
+        }
+
+        if (tokensToRefresh.length > 0) {
+          const installationUpdates: any = { ...queryResult }; // TODO :: TS
+          client = new WebClient(undefined, this.clientOptions);
+
+          // TODO :: Concurrent Calls?
+          // TODO :: Retry Strategy
+          for (const refreshToken of tokensToRefresh) {
+
+            const refreshResp = await client.oauth.v2.access({
+              client_id: this.clientId,
+              client_secret: this.clientSecret,
+              grant_type: 'refresh_token',
+              refresh_token: refreshToken,
+            }) as OAuthV2Response;
+
+            // TODO :: Sort out TS issues
```

Comment:
Indeed, as we use automatic code generator for the types, we are unable to have non-optional properties here. 
When we can assume the property exists 100%, using `!` can be okay but if I wrote this part, I would have if statements verifying the existence. Doing this satisfies the TS compiler plus may detect really unexpected situations (say, Slack server-side regressions / incidents).

```typescript
            const tokenType = refreshResp.token_type;
            if (tokenType === undefined) {
              throw new Error("unexpectedly token_type is missing here!");
            }
            const expireIn = refreshResp.expires_in;
            if (expireIn === undefined) {
              throw new Error("unexpectedly expire_in is missing here!");
            }
            const expiresAt: number = currentUTCSec + expireIn;

            if (tokenType === 'bot') {
              authResult.botToken = refreshResp.access_token;
              authResult.botRefreshToken = refreshResp.refresh_token;
              authResult.botTokenExpiresAt = expiresAt;
            }
```

Reply:
This type is actually not using the generated code version of the response, but it brings up the greater question: should it? Are we throwing out all of our hand-rolled types and opting for the generated ones at this point?

Thanks for your suggestion for a resolution the current state of things -- I'll plan to incorporate it.

Reply:
>This type is actually not using the generated code version of the response, but it brings up the greater question: should it?

Yes, the type will be updated once the properties are GAed (or at least in public beta). With the current approach, tall the properties in responses are optional. There is no way to exclude specific properties at this point. During the beta period, you can go with (=internally cast the type) an internal type in this source file.

>Are we throwing out all of our hand-rolled types and opting for the generated ones at this point?

No, we don't have the bandwidth to guarantee the quality of it for all API responses and I have to say that the situation can be the same in the future. That's why we are auto generating the code.

Reply:
For OAuth, we do have hand rolled types right now. See https://github.com/slackapi/node-slack-sdk/blob/main/packages/oauth/src/index.ts#L705-L726

But maybe we shouldn't have these and instead be switching over to using the generated types

Reply:
Every time we remove types that reliably indicate what properties are present I die a little inside, but I do acknowledge the motivation outlined above. 🙂  

Just let me know if you two would like me to remove it when the generated types are updated to reflect the payload changes!

Reply:
haha I agree. Lets not remove anything right now. Maybe in the future we can get more reliable with the generated response types and then we can revist this.